### PR TITLE
feat: Add `chat/messages` profile and slack map

### PR DESCRIPTION
**Pull request by martinalbert**
## Description
Adds profile for getting past messages in slack.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by janhalama** [comment]
@martinalbert is this ready for review? There are unfinished TODO tasks in the description.

**Comment by jnv** [comment]
I don't see this much as a blockers, Discord should be definitely in a separate PR, but if you have more tests in mind, it's better to do them now. :wink:

**Comment by martinalbert** [comment]
@janhalama @jnv it's ready, I removed discord from this PR and added missing things.

File: grid/chat/get-messages/profile.supr [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+model Message {
+	"""
+	Id
+	Identification of message
+	"""
+	id! string!
+
+	"""
+	Author
+	Author of retrieved message
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:75]: What should I expect here? An ID? A name? An avatar URL...?

File: grid/chat/get-messages/profile.supr [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"""
+	author! string!
+
+	"""
+	Destination
+	Destination of retrieved message
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:81]: Same as with input parameter. Also, do you need to return this if the user knows this upon input? :thinking:

File: grid/chat/get-messages/profile.supr [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	threadId string
+
+	"""
+	Text
+	Content of retrieved message
+	"""
+	text string
+
+	"""
+	CreatedAt
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:98]:
```suggestion
	Created At
```

File: grid/chat/get-messages/profile.supr [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	Date retrieved message was created
+	"""
+	createdAt string
+
+	"""
+	CreatedAt
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:104]:
```suggestion
	Updated At
```

File: grid/chat/get-messages/profile.supr [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+	"""
+	Reactions
+	List of reactions on retrieved message
+	"""
+	reactions [
+		{
+			emoji string
+			count number
+            users [string]
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:117]: Formatting? Also, it's not clear if this gives me a list of users' IDs or their names or whatever.

File: grid/chat/get-messages/profile.supr [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	"""
+	updatedAt string
+
+	"""
+	Reactions
+	List of reactions on retrieved message
+	"""
+	reactions [
+		{
+			emoji string
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:115]: I am uncertain about naming the property as "emoji", but since all platforms seem to name custom ones as "custom emojis", let's keep it. However, it should be specified what I get here; is it a text alias of the emoji like `:+1:` or `+1` or a unicode identifier?

Preferably mention this info the explanation of the `reactions` property (since I think we don't render nested properties well on the web).

File: grid/chat/get-messages/profile.supr [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+			count number
+            users [string]
+		}
+	]
+}
+
+model Messages {
+	"""
+	PreviousPage
+	Identification of previous page
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:125]:
```suggestion
	Cursor to the previous page for paginating backwards
```

File: grid/chat/get-messages/profile.supr [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	"""
+	previousPage string
+	
+	"""
+	NextPage
+	Identification of next page
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:131]: See my comments for previousPage

File: grid/chat/get-messages/profile.supr [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	nextPage string
+
+	"""
+	Messages
+	List of retrieved messages
+	"""
+	messages [Message]
+}
+
+model Error {
```
Comment (jnv) [inline grid/chat/get-messages/profile.supr:142]: Mind the formatting (we really should get Comlink into Prettier or something)

File: grid/chat/get-messages/maps/get-messages.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable jest/no-export */
+
+import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';
+
+export const getMessagesTest = (
```
Comment (jnv) [inline grid/chat/get-messages/maps/get-messages.ts:5]: I'd like to see pagination in action in this test (at least for n pages)

File: grid/chat/get-messages/maps/mock.suma [hunk 11]
```diff
@@ -0,0 +1,21 @@
+profile = "chat/get-messages@1.0"
+provider = "mock"
+
+"""
+GetMessages map
+"""
+map GetMessages {
```
Comment (jnv) [inline grid/chat/get-messages/maps/mock.suma:7]: Wow, this mock looks sad, but okay :joy:

**Review by jnv — CHANGES_REQUESTED** [review]

File: grid/chat/get-messages/profile.supr [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		AfterDate
+		Date that marks beginning of messages range until current time
+		"""
+		afterDate string
+        
+		"""
+		Limit
+		Maximum number of records to retrieve
+		"""
+		limit number
```
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:26]: It would be useful with slack where you can set up to 1k messages, where default is 100. I'll fix descriptions.

File: grid/chat/get-messages/profile.supr [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	input {
+		"""
+		Destination
+		Identification of destination to retrieve messages from
+		"""
+		destination! string!
+		
+		"""
+		AfterDate
+		Date that marks beginning of messages range until current time
```
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:18]: Slack expects UNIX timestamp and discord uses filtering based on message id, so we should probably support both. Discord also returns messages with ISO format timestamp.

File: grid/chat/get-messages/profile.supr (hunk 1 above)
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:75]: An ID, same with destination.

File: grid/chat/get-messages/profile.supr [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+	"""
+	Text
+	Content of retrieved message
+	"""
+	text string
+
+	"""
+	CreatedAt
+	Date retrieved message was created
```
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:99]: Good point, I'll make it required and fix description.

File: grid/chat/get-messages/profile.supr (hunk 5 above)
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:117]: I'll add description for it 👍🏻

File: grid/chat/get-messages/profile.supr (hunk 13 above)
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:18]: Going through discord docs, id of each message is snowflake, which contains timestamp, so maybe supporting UNIX only will be better. 
https://discord.com/developers/docs/reference#snowflakes
https://discord.com/developers/docs/resources/channel#message-object

File: grid/chat/get-messages/profile.supr (hunk 6 above)
Comment (martinalbert) [reply grid/chat/get-messages/profile.supr:115]: Slack returns identifier of emoji between colons, so in case of 👍🏻 it returns `+1`

**Comment by martinalbert** [comment]
@jnv I've added better description and mapping. Also, I had to go with UNIX timestamp as slack does not return ISO format for date of message creation. I'll add rate limit information and finish up discord in another PR.

File: grid/chat/get-messages/profile.supr [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+	nextPage string
+
+	"""
+	Messages
+	List of retrieved messages
+	"""
+	messages [Message]
+}
+
+model Error {
```
Comment (janhalama) [inline grid/chat/get-messages/profile.supr:143]: I miss rate limit in `Error` model.

File: grid/chat/get-messages/maps/slack.suma [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+		response 200 "application/json" {
+			// See https://api.slack.com/methods/conversations.history#errors for more information about errors
+			return map error if (body.ok === false) {
+				title = body.error
+			}
+
+			messages = call MapMessages(messages = body.messages)
+
+			return map result if (body.has_more) {
```
Comment (janhalama) [inline grid/chat/get-messages/maps/slack.suma:30]: Looks like Slack dosn't support fetching previous page.

Comment (martinalbert) [reply grid/chat/get-messages/maps/slack.suma:30]: Yeah, I didn't find a way to link previous pages with slack. 🤷🏻‍♂️

**Review by jnv — CHANGES_REQUESTED** [review]
I think some of my comments were not addressed yet? :thinking:

File: grid/chat/get-messages/maps/slack.suma [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+				title = body.error
+			}
+		}
+	}
+}
+
+operation MapMessages {
+	messages = args.messages.filter(message => message.type === 'message')
+					
+	set if (messages.length > 0) {
```
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/maps/slack.suma:58]: What do you mean?

[the remark this answers is not in the document]
Comment (jnv) [reply grid/chat/get-messages/maps/slack.suma]: What I mean is that you could remove that condition completely. Or just chain it into a single call:

```js
args.messages.filter(message => message.type === 'message').map(() => ...)
```

Though more efficient approach to avoid 2 separate iterations would be to do filtering in single pass, e.g.

```js
messages = []
args.messages.forEach((message) => {
  if (message.type !== 'message') {
    return
  }
  // ...
  messages.push(result)
}
```

File: grid/chat/messages/profile.supr [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+		Destination
+		Representation of a messaging destination. Can be ID or channel name, depending on the provider capability.
+		"""
+		destination! string!
+		
+		"""
+		After Date
+		Retrieve only messages after this date and time, UNIX timestamp format (e.g. 1095379198.75) 
+		"""
+		afterDate number
```
Comment (jnv) [inline grid/chat/messages/profile.supr:20]: I think we discussed that calling it Date is misleading since it expects a timestamp, no?

```suggestion
		"""
		After Timestamp
		Retrieve only messages after the given UNIX timestamp (in seconds, e.g. 1095379198.75) 
		"""
		afterTimestamp number
```

File: grid/chat/messages/profile.supr [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+		
+		"""
+		Before Date
+		Retrieve only messages before this date and time, UNIX timestamp format (e.g. 1095379198.75)
+		"""
+		beforeDate number
```
Comment (jnv) [inline grid/chat/messages/profile.supr:26]: Same as with afterDate

File: grid/chat/messages/profile.supr [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+		}
+	}
+}
+
+model Attachment {
+	"""
+	Id
+	Unique identifier of the attachment
+	"""
+	id! string!
```
Comment (jnv) [inline grid/chat/messages/profile.supr:90]: I wouldn't require an attachment's ID as mandatory field.

File: grid/chat/messages/profile.supr [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+
+	"""
+	Created At
+	UNIX timestamp (e.g. 1095379198.75) when the attachment was created
+	"""
+	createdAt! number!
```
Comment (jnv) [inline grid/chat/messages/profile.supr:96]: I am somewhat torn here, we use `createdAt` for ISO date fields in other profiles, so maybe use `createdTimestamp` instead?

File: grid/chat/messages/profile.supr [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+
+	"""
+	File Name
+	Name of attached file
+	"""
+	fileName string
+
+	"""
+	Media Type
+	Format of the attached file
```
Comment (jnv) [inline grid/chat/messages/profile.supr:106]: Would be useful to mention what to expect here; MIME type? String like "video", "image"? Or it's provider specific?

File: grid/chat/messages/profile.supr [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+
+	"""
+	Url
+	Source url of the attachment
+	"""
+	url string
+
+	"""
+	Preview
+	Preview of the attached file
```
Comment (jnv) [inline grid/chat/messages/profile.supr:118]: What is this? An URL with the preview? Is it preview image? If this is the case, I suggest renaming this to `previewUrl`

File: grid/chat/messages/profile.supr [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+	author! {
+		id! string!
+		username string!
+	}!
+
+	"""
+	Created At
+	UNIX timestamp (e.g. 1095379198.75) when the message was created
+	"""
+	createdAt! number!
```
Comment (jnv) [inline grid/chat/messages/profile.supr:143]: Again with createdAt :thinking:

File: grid/chat/messages/profile.supr [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+	Text
+	Contents of the retrieved message
+	"""
+	text string
+
+	"""
+	Updated At
+	UNIX timestamp (e.g. 1095379198.75) when the message was last updated
+	"""
+	updatedAt number
```
Comment (jnv) [inline grid/chat/messages/profile.supr:173]: updatedTimestamp?

File: grid/chat/messages/maps/slack.suma [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	return details
+}
+
+operation ComposeURL {
+	timestamp = args.ts
+	destination = args.dest
+
+	return `https://slack.com/api/conversations.history?channel=${destination}&latest=${timestamp}&inclusive=true&limit=1`
```
Comment (jnv) [inline grid/chat/messages/maps/slack.suma:175]: Huh, if this is publicly exposed URL, I don't think it should point to an API, but maybe I misunderstand how this is used?

**Review by jnv — CHANGES_REQUESTED** [review]
A few notes regarding docstring.

I am a bit torn on `*At` and `*Timestamp` but I am probably okay with using `createdAt` and `updatedAt` with timestamps.

File: grid/chat/messages/profile.supr (hunk 18 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:20]: This is implemented in PR with discord, I'll add it here as well.

File: grid/chat/messages/profile.supr (hunk 20 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:90]: Both providers return id of attachments 🤷🏻‍♂️

File: grid/chat/messages/profile.supr (hunk 21 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:96]: I think suffix `At` is general enough to be referencing to timestamp as well.

File: grid/chat/messages/profile.supr (hunk 23 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:118]: This only applies to slack and it contains string preview of attachment file.

File: grid/chat/messages/maps/slack.suma (hunk 26 above)
Comment (martinalbert) [reply grid/chat/messages/maps/slack.suma:175]: Yeah, this points to the API 😕

File: grid/chat/messages/profile.supr (hunk 21 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:96]: It's also used in other capabilities in station.

File: grid/chat/messages/profile.supr (hunk 20 above)
Comment (jnv) [reply grid/chat/messages/profile.supr:90]: Hmm, okay, so keep it, perhaps we will need it for some consecutive use-cases.

File: grid/chat/messages/profile.supr (hunk 23 above)
Comment (jnv) [reply grid/chat/messages/profile.supr:118]: Okay, so looking at docs: https://api.slack.com/types/file#previews
Seems this has limited usage for text snippets. I wouldn't mind removing it. Alternatively please expand the description like "Short text preview of text attachments (e.g. snippets, posts)"

Comment (martinalbert) [reply grid/chat/messages/profile.supr:118]: I agree, I'll remove it.

File: grid/chat/get-messages/maps/slack.suma (hunk 17 above)
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/maps/slack.suma]: That looks good, I'll refactor it.

File: grid/chat/messages/profile.supr (hunk 22 above)
Comment (martinalbert) [reply grid/chat/messages/profile.supr:106]: It's MIME type in both providers, will this be enough? 👇🏻 

```suggestion
	MIME type of the attached file.
```

File: grid/chat/get-messages/maps/slack.suma (hunk 17 above)
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/maps/slack.suma]: Unfortunately jessie expression such as `args.messages.forEach` can't be used outside assignment, so I had to chain it, but when I chaned it, language server had problem with that, so I just removed the condition.

[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/get-messages/maps/slack.suma]: Ok it no longer throws that error, so I chained it.

**Comment by martinalbert** [comment]
@janhalama @jnv I added some review fixes and also refactored mapping due to performance issues.

File: grid/chat/messages/profile.supr (hunk 22 above)
Comment (jnv) [reply grid/chat/messages/profile.supr]: Great, thanks!

File: package.json [hunk 27]
```diff
@@ -22,7 +22,7 @@
   },
   "devDependencies": {
     "@superfaceai/ast": "^1.0.0",
-    "@superfaceai/cli": "^1.0.1",
+    "@superfaceai/cli": "1.1.0-beta.0",
```
Comment (jnv) [inline package.json:25]: Is this change necessary for this PR, can't it be done in a separate PR?

File: grid/chat/messages/maps/slack.suma [hunk 28]
```diff
@@ -0,0 +1,230 @@
...
+        fileName: file.name,
+        mediaType: file.mimetype,
+        url: file.permalink_public,
+        preview: file.preview,
+      }));
+
+      resultMessage.attachments = attachments;
+    }
+
+    resultMessage.url = `https://slack.com/api/conversations.history?channel=${args.destination}&latest=${message.ts}&inclusive=true&limit=1`;
```
Comment (jnv) [inline grid/chat/messages/maps/slack.suma:116]: Okay, so this is last outstanding issue for me.

I see that Slack has a dedicated method for that, which would be costly to run for every message: https://api.slack.com/methods/chat.getPermalink

But I'd say consumer's expectation here is to return a URL which is presentable to the end user and take them to the relevant message. So returning this is in my opinion worse than returning nothing at all.

What is missing to be able to construct this URL manually. For example here is threaded message:

```
https://superfaceai.slack.com/archives/C013602CL2Y/p1645175948429299?thread_ts=1645043940.215569&cid=C013602CL2Y
```

I don't think we have any workspace identifier (or do we?), but we do have channel ID, message ID and thread ID.

**Review by jnv — CHANGES_REQUESTED** [review]
Message url is a blocker for me.

File: package.json (hunk 27 above)
Comment (martinalbert) [reply package.json:25]: It's mainly to unblock lint in CI 🤷🏻‍♂️

File: grid/chat/messages/maps/slack.suma (hunk 28 above)
Comment (martinalbert) [reply grid/chat/messages/maps/slack.suma:116]: I'll try to create public URL then. 🙏🏻

File: grid/chat/messages/maps/slack.suma [hunk 29]
```diff
@@ -0,0 +1,231 @@
+profile = "chat/messages@1.0"
+provider = "slack"
+
+"""
+GetMessages map
+"""
+map GetMessages {
+	http GET "/conversations.history" {
```
[the remark this answers is not in the document]
Comment (martinalbert) [reply grid/chat/messages/maps/slack.suma]: I updated formatting in all PRs to 2 space indentation.

File: grid/chat/messages/maps/slack.suma (hunk 28 above)
Comment (martinalbert) [reply grid/chat/messages/maps/slack.suma:116]: I would also need workspace name to compose correct URL, so it's not possible without that call.

We could return partially generated url, user would have to add his workspace name. For example, I'm able to generate URL: `https://slack.com/archives/CF3H7S63W/p1644310634546009`, but to make it work, I have to add workspace name before domain i.e. `https://<workspace-name>.slack.com/archives/CF3H7S63W/p1644310634546009`. URL for threads are more complicated, but still need workspace name.

**Comment by jnv** [comment]
Ok to go, let's rather add URLs in a separate PR.
